### PR TITLE
Add FROMADDR= parameters to SMTP SCHK command for JMAP future delivery

### PR DESCRIPTION
**Pull request by ksmurchison**

### CODE REVIEW

File: imap/smtpclient.c [hunk 1]
```diff
@@ -394,12 +394,25 @@ static int smtpclient_rset(smtpclient_t *sm)
     return r;
 }
 
-static int smtpclient_schk(smtpclient_t *sm)
+static int smtpclient_schk(smtpclient_t *sm, json_t *fromaddr)
```
Comment (rsto) [inline imap/smtpclient.c:397]: Is this the first time we pull in json into smtplib? If so, could we reuse the existing smtp_addr_t type (or a list of it) instead of a JMAP EmailAddress object?

File: imap/smtpclient.c [hunk 2]
```diff
@@ -863,6 +891,19 @@ EXPORTED int smtp_is_valid_esmtp_value(const char *val)
     return 1;
 }
 
+/* Encodes into the current location of the struct buf.
+   The caller must buf_reset() before calling is necessary. */
```
Comment (rsto) [inline imap/smtpclient.c:895]: Nit type: 'is' should mean 'if'. I am generally bad with comma, but I think the "if necessary" should be preceded by one.

File: imap/smtpclient.h [hunk 3]
```diff
@@ -108,6 +112,11 @@ extern int smtpclient_open_host(const char *addr, smtpclient_t **smp);
 extern int smtpclient_send(smtpclient_t *sm, smtp_envelope_t *env, struct buf *data);
 extern int smtpclient_sendprot(smtpclient_t *sm, smtp_envelope_t *env, struct protstream *data);
 
+/* Perform SMTP envelope (and optionally size and/or From: header address checks)
```
Comment (rsto) [inline imap/smtpclient.h:115]: This is find hard to parse. Would "Check the SMTP envelope" work instead of "Perform .."

**Review by rsto — COMMENTED** [review]
Logic looks good to me. Found some nits.

File: imap/smtpclient.c (hunk 1 above)
Comment (ksmurchison) [reply imap/smtpclient.c:397]: I'm not too keen on passing json into smtpclient.c either but I also didn't want to have to translate the json array into some other format, but its probably best that I do.

Comment (ksmurchison) [reply imap/smtpclient.c]: Changed to using a strarray_t

**Review by rsto — APPROVED** [review]
Looks good to me
